doc/development/guide/documentation.rst

```diff
@@ -1,39 +1,455 @@
+
+.. code-block:: python
+
+    x = {"John": 23, "James": 54}
+    """dict[str, int]: stores the ages of known users"""
+
+The syntax is the same as those used for describing function arguments.
+
+
+.. _code_examples:
+
+Code examples
+-------------
+
+Code examples are very important; they *show* readers how the function or class should be used.
+When writing code examples for docstrings, use the following guidelines:
+
+- You may assume that PennyLane is imported as ``qml`` and NumPy is imported as ``np`` in the code examples.
+  All other imports must be specified explicitly.
+
+- For single line statements and associated output, use Python console syntax (``pycon``):
+
+  .. code-block:: rest
+
+      >>> circuit(0.5, 0.1)
+      [0.43241234312, -0.543534534]
+
+  Multi-line statements should use ``...`` to indicate continuation lines:
+
+  .. code-block:: rest
+
+      >>> dev = qml.device("default.qubit", wires=1)
+      >>> @qml.qnode(dev)
+      >>> def circuit(x):
+      ...     qml.RX(x, wires=0)
+      ...     return qml.expval(qml.PauliZ(0))
+      >>> circuit(0.5)
+      0.8775825618903726
+
+  For larger, more complicated code blocks, favour standard Python code-block with
+  Python console syntax for displaying output:
+
+  .. code-block:: rest
+
+      .. code-block:: python3
+
+          dev = qml.device("default.qubit", wires=1)
+          @qml.qnode(dev)
+          def circuit(x):
+              qml.RX(x, wires=0)
+              return qml.expval(qml.PauliZ(0))
+
+      Executing this circuit:
+
+      >>> circuit(0.5)
+      0.8775825618903726
+
+
+Commenting code
+---------------
+
+While docstrings describe *what* a function or class does, or how it is used, **comments**
+are used to document the underlying implementation. Use comments where applicable to make
+your code easy to follow and understand, keeping to the following guidelines.
+
+* **Use comments to explain the implementation or algorithm, never to describe the code.**
+  Assume the reader has basic understanding of common programming principles and syntax.
+  Do not assume the reader knows what you're trying to do with it!
+
+  .. note::
+
+      If additional implementation details have been requested by reviewers during code review,
+      these must be incorporated as code comments.
+
+* **Code should be self-documented where possible.** Code should be clear and concise,
+  with the logical flow easily followed by the reader. This can be done by using common
+  programming patterns alongside sensible variable and function names.
+
+  Note that this does not exclude the use of code comments; comments are particularly valuable
+  in quantum software, where the underlying algorithm can be highly non-trivial. Rather,
+  self-documented code simply excludes *unnecessary* comments.
+
+* **Comments should be as close to the described code as possible.** These can either be
+  single- or multi-line comments *above* the described code block,
```

Comment:
Should a single line comment have an empty line between the comment or the line as in the example below, or should it actually be right above (as close as possible)?

Reply:
I'm happy to leave this up to the code author/code review. But I think you're right, this particular example is confusing, so I've removed the line break.